### PR TITLE
HTTP2: Add graceful goaway

### LINKED ISSUE #24237: Server does not use graceful GOAWAY on graceful shutdown
The server seems to have a concept for graceful shutdown, where it sends a GOAWAY and lets RPCs complete. However, it doesn't do the double-GOAWAY suggested by [MAX_CONNECTION_AGE gRFC](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md#proposal) and the [HTTP/2 spec](https://httpwg.org/specs/rfc7540.html#GOAWAY):

> A server that is attempting to gracefully shut down a connection SHOULD send an initial GOAWAY frame with the last stream identifier set to 2<sup>31</sup>-1 and a NO_ERROR code. This signals to the client that a shutdown is imminent and that initiating further requests is prohibited. After allowing time for any in-flight stream creation (at least one round-trip time), the server can send another GOAWAY frame with an updated last stream identifier. This ensures that a connection can be cleanly shut down without losing requests.

I'd suggest an implementation that sends a PING immediately after the first GOAWAY and sends the second GOAWAY when the ACK arrives. A timer can be set to 10 seconds and the second GOAWAY can be sent when the timer expires if the ACK has failed to arrive. This allows rapid, zero-configuration, race-less shutdowns.

The current server behavior is causing unnecessary RPC failures as it fails RPCs already sent by the client. This has been seen internally when a client swapped the server from GFE to C core, but also in https://github.com/grpc/grpc-java/issues/7409 . I believe we haven't noticed this before because many clients are automatically retrying these failed RPCs, but a bug in grpc-java (https://github.com/grpc/grpc-java/issues/7357) prevented them from showing up as UNAVAILABLE. gRPC clients have also have local races relating to GOAWAY and so it can been hard to determine whether seen instances are client-side races. The internal case was also easier to pin-point because we could do A/B comparison between GFE and C core.

MAX_CONNECTION_AGE is probably more important than normal server shutdown, as it occurs more frequently. But normal server shutdown should also avoid this race. [Go](https://github.com/grpc/grpc-go/blob/8630cac324bf02ea0edfba758c2b2f3344af7bf7/internal/transport/http2_server.go#L1191), [Java](https://github.com/grpc/grpc-java/blob/c29ad76daea0d17637fefcab4dbfd9ef8cb84ebc/netty/src/main/java/io/grpc/netty/NettyServerHandler.java#L939), and GFE have had this behavior for years.

CC @veblush

### PR DESCRIPTION
**Pull request by yashykt**
Fixes #24237
Implements graceful goaway at the HTTP2 layer.

Before this PR, the behavior of the HTTP2 server on shutdown was - 
* Send a GOAWAY frame with last stream ID filled with the last stream ID that was accepted.
* Keep the connection alive till the number of streams goes down to 0.
* Additionally, it would keep accepting new streams even after sending this GOAWAY frame. This is broken behavior as per the HTTP2 spec.

After this PR, the HTTP2 server would perform a graceful shutdown as recommended by the spec -
* Send a GOAWAY with last stream ID as 2^31-1 indicating to the client that no new streams should be started.
* Waits for an RTT period by sending a ping (and waiting for an ACK). A hard limit of 20 seconds is considered as the limit beyond which we don't wait for the ACK anymore. While this is in progress, we continue to accept new streams to handle streams that were started before the client saw the GOAWAY.
* After the RTT, we send a GOAWAY with last stream ID as the last stream that was accepted till this point. No more new streams are accepted after this.
* The connection then remains alive till the existing streams are finished.

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 1]
```diff
@@ -1759,18 +1759,123 @@ void grpc_chttp2_ack_ping(grpc_chttp2_transport* t, uint64_t id) {
...
+namespace {
+
+// Fire and forget (deletes itself on completion). Does a graceful shutdown by
+// sending a GOAWAY frame with the last stream id set to 2^31-1, sending a ping
+// and waiting for an ack (effective waiting for an RTT) and then sending a
+// final GOAWAY freame with an updated last stream identifier. This helps ensure
+// that a connection can be cleanly shut down without losing requests.
+// In the event, that the client does not respond to the ping for some reason,
+// we add a 20 second deadline, after which we send the second goaway.
+class GracefulGoaway : public grpc_core::InternallyRefCounted<GracefulGoaway> {
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1771]: let's just make this RefCounted<> - we're not using the Orphan feature, and nothing else can form a pointer to this type, so no need to try and hide the Ref/Unref.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 2]
```diff
@@ -1759,18 +1759,123 @@ void grpc_chttp2_ack_ping(grpc_chttp2_transport* t, uint64_t id) {
...
+        &timer_,
+        grpc_core::ExecCtx::Get()->Now() + grpc_core::Duration::Seconds(20),
+        GRPC_CLOSURE_INIT(&on_timer_, OnTimer, this, nullptr));
+  }
+
+  // Unused
+  void Orphan() override { GPR_ASSERT(0); }
+
+  void MaybeSendFinalGoawayLocked() {
+    if (t_->sent_goaway_state != GRPC_CHTTP2_GRACEFUL_GOAWAY) {
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1798]: Should this be an assert?

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 3]
```diff
@@ -1759,18 +1759,123 @@ void grpc_chttp2_ack_ping(grpc_chttp2_transport* t, uint64_t id) {
   }
 }
 
+namespace {
+
+// Fire and forget (deletes itself on completion). Does a graceful shutdown by
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1764]: My understanding of this type is that it keeps the transport alive an extra 20 seconds unconditionally - I think this is problematic -- if we signal GOAWAY and wait one RTT, then once the number of requests on a transport drops to zero we are able to complete this dance deterministically and release the (significant) memory used by the transport much earlier.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc (hunk 1 above)
Comment (ctiller) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1771]: Consider RefCounted<> here - we don't need the additional InternallyRefCounted machinery.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc (hunk 3 above)
Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1764]: No, it's not an unconditional 20 second wait. It only waits 20 seconds if the peer does not respond to the ping. If there are no streams and the RTT completes with a ping ack, then we would immediately send the final goaway and close the transport. (Some of our tests fail on an unconditional wait.)

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc (hunk 2 above)
Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1798]: No, this handles a race condition between the timer and the ping ack. Consider a scenario where the timer fires but we also receive the ping ack around the same time.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc (hunk 3 above)
Comment (ctiller) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1764]: I missed the grpc_timer_cancel last night! That makes sense... let me take another pass on this.

**Comment by yashykt** [comment]
Thanks for reviewing!

(Ignoring the C# infra flake - ```unzip:  cannot find or open ../../../../input_artifacts/csharp_nugets_windows_dotnetcli.zip, ../../../../input_artifacts/csharp_nugets_windows_dotnetcli.zip.zip or ../../../../input_artifacts/csharp_nugets_windows_dotnetcli.zip.ZIP.
```)

**Pull request merged**